src/common/gui/base/Toggle.ts

```diff
@@ -0,0 +1,98 @@
+import m, { ClassComponent, Vnode } from "mithril"
+import { Keys, TabIndex } from "../../api/common/TutanotaConstants.js"
+import { isKeyPressed } from "../../misc/KeyManager.js"
+
+type ToggleVariant = "normal" | "expanded"
+type TogglePosition = "left" | "right"
+type HTMLElementWithAttrs = Partial<Omit<HTMLElement, "style"> & ToggleAttrs>
+
+export interface ToggleAttrs {
+	checked: boolean
+	onclick: (checked: boolean) => unknown
+	ariaLabel: string
+	disabled?: boolean
+	togglePosition?: TogglePosition
+	class?: string
+	variant?: ToggleVariant
+}
+
+/**
+ * Toggle component with variants
+ * @see Component attributes: {ToggleAttrs}
+ * @example
+ * m(Toggle,
+ *     {
+ *         checked: this.checked,
+ *         onclick: (checked: boolean) => {
+ *             this.checked = checked
+ *             console.log(this.checked)
+ *         },
+ *         togglePosition: "right",
+ *         ariaLabel: "Test Toggle",
+ *         disabled: false,
+ *         variant: "normal",
+ *     },
+ *     "My label",
+ * ),
+ */
+
+export class Toggle implements ClassComponent<ToggleAttrs> {
+	private checkboxDom?: HTMLInputElement
+
+	view({ attrs: { disabled, variant, ariaLabel, checked, onclick, togglePosition }, children }: Vnode<ToggleAttrs>) {
+		const childrenArr = [children, this.buildTogglePillComponent(onclick, disabled)]
+		if (togglePosition === "left") {
+			childrenArr.reverse()
+		}
+
+		return m(
+			"label.toggle",
+			{
+				class: this.resolveClasses(disabled, variant),
+				role: "checkbox",
+				ariaLabel: ariaLabel,
+				ariaChecked: String(checked),
+				ariaDisabled: disabled ? "true" : undefined,
+				tabIndex: Number(disabled ? TabIndex.Programmatic : TabIndex.Default),
+				onkeydown: (e: KeyboardEvent) => {
+					if (isKeyPressed(e.key, Keys.SPACE, Keys.RETURN)) {
+						e.preventDefault()
+						this.checkboxDom?.click()
+					}
+				},
+			} satisfies HTMLElementWithAttrs,
+			childrenArr,
+		)
+	}
+
+	private buildTogglePillComponent(onclick: (checked: boolean) => unknown, disabled: boolean | undefined) {
+		return m(
+			"span.toggle-pill",
+			{
+				class: this.checkboxDom?.checked ? "checked" : "unchecked",
+			},
+			m("input[type='checkbox']", {
+				onclick: () => {
+					onclick(this.checkboxDom?.checked ?? false)
+				},
+				oncreate: ({ dom }) => {
+					this.checkboxDom = dom as HTMLInputElement
+				},
+				tabIndex: TabIndex.Programmatic,
```

Comment:
Just for clarity, we set the checkbox's tab index to unavailable, but we can tab to the parent and do the logic there, right?

Reply:
Yes, the checkbox is to transfer the state update to the browser. Then we just listen to changes there